### ISSUE TITLE
3000x slow down for outer indexing between netCDF 4.6.1 and 4.6.2

### ISSUE THREAD

**Issue by shoyer** [Unidata/netcdf4-python#870]
A xarray user reported a 1000 fold slow down for some operations between netCDF-C 4.6.1 and 4.6.2: https://github.com/pydata/xarray/issues/2747

The culprit appears to be large "outer indexing" operations. Consider the following netCDF file and script (e.g., run in IPython):
[test.nc.zip](https://github.com/Unidata/netcdf4-python/files/2834623/test.nc.zip)
```
import netCDF4
import numpy as np
f = netCDF4.Dataset('test.nc')
%time f['foo'][np.arange(1000), np.arange(100)]
```

On my laptop, this takes 2.04 ms using libnetcdf 4.6.1, but 5.94 seconds using libnetcdf 4.6.2.

Looking at the upstream release notes, it appears the culprit is most likely the indexing refactoring from https://github.com/Unidata/netcdf-c/pull/1001.

**Comment by DennisHeimbigner** [comment]
Can someone provide an equivalent in pseudocode of this:
>f['foo'][np.arange(1000), np.arange(100)]
to the level of nc_get_vars?

**Comment by shoyer** [comment]
It looks like the issue is that netCDF4-Python has special case logic (based on the libnetcdf version) for whether it sets `_no_get_vars` on variables:
https://github.com/Unidata/netcdf4-python/blob/a5325ff281d3ce178262eb135eca5751b2f4cd29/netCDF4/_netCDF4.pyx#L3805-L3813

If `_no_get_vars` is True (only for old libnetcdf), then netCDF4-Python will try to apply an optimization that replaces repeated indexing (100,000 calls to `nc_get_vars` in this case!) with a single call, when possible:
https://github.com/Unidata/netcdf4-python/blob/master/netCDF4/utils.py#L260-L263

So the simple fix might be to re-enable this optimization in netCDF4-Python, even with the newer version of libnetcdf.

**Comment by shoyer** [comment]
It was definitely a false alarm on [Unidata/netcdf-c#1001](https://github.com/Unidata/netcdf-c/pull/1001. The speed seems to be entirely dependent on the value of the `_no_get_vars` flag in netCDF4-Python.

**Comment by jswhit** [comment]
There was a flaw in the no_get_vars logic in utils.py.  Fixed by pull request #872.

**Comment by jswhit** [comment]
pull request #872 merged, this should be fixed now in master.  Closing.
